Studio inspired fixes

davidtheclark opened the pull request:
- Add `Draw.getMode()` API function, so Studio could tell if it needed to update Draw's mode.
- Make `Draw.set()` re-render when it's done deleting.
- Prevent box selecting from happening because you shift+double-clicked to zoom out
- Exit `direct_select` mode if the selected feature has been deleted
- Add `Draw.setSelected()` to API as a preferable way to select items programmatically, instead of needing to change mode to `simple_select` and specifying features.
- Fix cursors so the right ones show and they change as the mouse moves.
- Fix this: create a feature, then before deselecting it try to drag it — it's like the browser tries to treat it as a file that you're dragging and dropping. Fixed by re-rendering the feature when you mousedown on it.

Might accumulate more ...

davidtheclark commented:
Working on this in parallel with https://github.com/mapbox/studio/pull/7087

mcwhittemore commented:
> Make Draw.set() re-render when it's done.

This makes me thing that Draw.add or Draw.delete isn't rerendering as we'd like it to...

davidtheclark commented:
More:
- Prevent box selecting from happening because you shift+double-clicked to zoom out
- Exit `direct_select` mode if the selected feature has been deleted

davidtheclark commented:
> This makes me thing that Draw.add or Draw.delete isn't rerendering as we'd like it to...

@mcwhittemore True, I'll look into that ...

davidtheclark commented:
@mcwhittemore I think the issue is that at https://github.com/mapbox/mapbox-gl-draw/blob/studio-inspired-fixes/src/api.js#L42 `Draw.set()` is using the API's `add()` function, which does re-render when called, but at https://github.com/mapbox/mapbox-gl-draw/blob/studio-inspired-fixes/src/api.js#L52 it's using the store's `delete()` function, which does not re-render when called. I think it actually makes sense to use the store's methods in both places then save the rendering for the end of `Draw.set()`, right? To avoid rendering once for add and once for delete? Do you agree?

davidtheclark commented:
@mcwhittemore Nevermind ... I see that there's much more going on in API's `add` than in Store's, and `Draw.set()` needs that. I'll just use API's methods for both, for now.

mcwhittemore commented:
@davidtheclark I've tried to keep raw GeoJSON out of the store as when we use the store internally (in the `draw_*` modes) we want to add feature objects rather than geojson.

If the `Draw.delete()` took an array of feature ids we could use it in `Draw.set()` and get around this.

davidtheclark commented:
- Adds `Draw.setSelected()` to API as a preferable way to select items programmatically, instead of needing to change mode to `simple_select` and specifying features.

davidtheclark commented:
@mcwhittemore This is ready for review. I tried to put all the changes in a list in the first comment of this thread. Once we merge this into master, let's double-check docs, create issues for necessary tests, and try to get a release out.

mcwhittemore commented on src/api.js:
```diff
@@ -82,8 +94,8 @@ module.exports = function(ctx) {
     },
-    delete: function(id) {
-      ctx.store.delete([id], { silent: true });
+    delete: function(featureIds) {
```
This is a breaking change. We need to update the docs.

davidtheclark replied:
It is actually not breaking: it will accept either a single ID, as before, or an array. But, yes, we need to clean up docs with these changes.

mcwhittemore commented on src/api.js:
```diff
@@ -23,6 +25,15 @@ module.exports = function(ctx) {
     getSelectedIds: function () {
       return ctx.store.getSelectedIds();
     },
+    setSelected: function(idsToSelect) {
```
This is duplicate functionality to `Draw.changeMode()` that automatically changes the mode for you. Once we can do this without changing the mode I think its great, but until then lets just use what we have.

mcwhittemore commented on src/modes/simple_select.js:
```diff
@@ -75,6 +79,15 @@ module.exports = function(ctx, options = {}) {
 
+      // Any click should stop box selecting and dragging
+      this.on('click', () => true, function() {
+        dragging = false;
```
Should we move these to `cleanupBoxSelect` as well?

mcwhittemore commented on src/lib/mode_handler.js:
```diff
@@ -37,15 +37,18 @@ var ModeHandler = function(mode, DrawContext) {
       }
     }
-    DrawContext.ui.updateMapClasses();
+    if (handlerCalled) {
```
I'm pretty sure this check is what caused the cursors to go away. Looks like you've fixed that problem by calling render a few times inside of the modes. Either way is good with me as long as the way we go is fastest :).

mcwhittemore commented on src/render.js:
```diff
@@ -6,9 +6,15 @@ module.exports = function render() {
+
+  // If we were in direct select mode and our selected feature no longer exists
+  // (because it was deleted), we need to get out of that mode.
+  if (mode === Constants.modes.DIRECT_SELECT && !store.get(store.getSelectedIds()[0])) {
```
hmmm... this is ok for now, but I'd like to find a way to keep mode logic out of the renderer. Maybe there is a way to emit events only to the modes when its an application triggered event. This way modes could listen to the delete and selection change events and update themselves.

mcwhittemore commented:
@davidtheclark  Functionality wise this is looking good. Mouse cursors all seem right and moving between modes seems right. I haven't checked the events other than looking at the tests.

davidtheclark commented:
@mcwhittemore I think I addressed your concerns:

- Removed `Draw.setSelected`.
- Moved the direct-select-escape-hatch out of `render` (into the API functions `delete` and `deleteAll`).

Ready for another review or merge.

mcwhittemore commented:
@davidtheclark merged and published at 0.11.0.
